### DISCUSSION TITLE
half integer genus

### DISCUSSION THREAD

**Discussion by rob-rb** [meshpro/meshplex#157] [General]
I often get half-integer genus for a surface with 3d-coordinates with boundaries. This is not an issue of meshplex i get this strange behaviour with other libraries as well. Does somebody how this can be explained?

**Comment by nschloe** [comment]
It's half-integer if `len(points) - len(edges) + len(cells)` is odd. I can't think of a situation where this could be true, but perhaps some half-overlapping cells? In any case, it should be "weird".

**Comment by rob-rb** [comment]
Actually for a mesh with boundary-edges there is a different formula for the genus:
it is g = (1- (V+F-E - b)/2) where b is the number of connected boundary components.

**Reply by nschloe** [reply]
Interesting! Source? Should be added to https://math.stackexchange.com/a/85164/36678.

**Comment by rob-rb** [comment]
source: https://en.wikipedia.org/wiki/Genus_(mathematics)
I have written the following code maybe you could add it to meshplex allthough you'll wont admire the greedy style:
```

def remove_connected_component(ht, keys, i, res):
    for k in ht[i]:
        if k in keys:
            keys.remove(k)
            res.append(k)
            remove_connected_component(ht, keys, k, res)


def connected_boundary_components(mesh):
    edges = mesh.edges["points"]
    bedges = list(edges[mesh.is_boundary_edge])
    ht = defaultdict(list)
    for i, j in bedges:
        ht[i].append(j)
        ht[j].append(i)
    keys = list(ht.keys())
    n = 0

    while keys:
        i = keys.pop(0)
        n += 1
        res = [i]
        remove_connected_component(ht, keys, i, res)
        
    return n



def genus_with_boundary(mesh):
    res = mesh.genus
    bc = optimesh_smooth.connected_boundary_components(mesh)
    res += bc*0.5
    return res
```

**Comment by nschloe** [comment]
Fixed in https://github.com/nschloe/meshplex/pull/158 now using [`connected_components`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csgraph.connected_components.html) from scipy.

**Reply by rob-rb** [reply]
wow looks elegant and fast. guess that möbius strip would have had half-integer genus with old code.

**Reply by nschloe** [reply]
Well, Möbius strips aren't orientable so even now we don't compute the genus for it. It correctly throws an error now, though, and should work for your sponge-like meshes.
